lib/decompress_impl.ml

```diff
@@ -1538,6 +1555,350 @@ module Zlib_deflate = struct
+    | Exception e -> Format.fprintf fmt "(Exception %a)" pp_error e
+
+  let pp fmt {d; z; _} =
+    Format.fprintf fmt "{@[<hov>d = @[<hov>%a@];@ z = %a;@]}"
+      RFC1951_deflate.pp d pp_state z
+
+  let ok t : ('i, 'o) res = Ok {t with z= Finish}
+  let error t exn : ('i, 'o) res = Error ({t with z= Exception exn}, exn)
+
+  let rec put_byte ~ctor byte k src dst t =
+    if t.d.RFC1951_deflate.o_len - t.d.RFC1951_deflate.o_pos > 0 then (
+      Safe.set t.d.RFC1951_deflate.wo dst
+        (t.d.RFC1951_deflate.o_off + t.d.RFC1951_deflate.o_pos)
+        (Char.unsafe_chr byte) ;
+      k src dst
+        { t with
+          d=
+            { t.d with
+              RFC1951_deflate.o_pos= t.d.RFC1951_deflate.o_pos + 1
+            ; RFC1951_deflate.write= t.d.RFC1951_deflate.write + 1 } } )
+    else
+      Flush
+        { t with
+          z=
+            ctor (fun src dst t ->
+                (put_byte [@tailcall]) ~ctor byte k src dst t ) }
+
+  let put_short_lsb ~ctor short k src dst t =
+    let put_byte = put_byte ~ctor in
+    (put_byte (short land 0xFF) @@ put_byte ((short lsr 8) land 0xFF) k)
+      src dst t
+
+  let put_string ~ctor str k src dst t =
+    let len = String.length str in
+    let string = Safe.of_string str in
+    let rec go ~ctor off src dst t =
+      let to_blit =
+        min (len - off) (t.d.RFC1951_deflate.o_len - t.d.RFC1951_deflate.o_pos)
+      in
+      Safe.blit_string t.d.RFC1951_deflate.wo string off dst
+        (t.d.RFC1951_deflate.o_off + t.d.RFC1951_deflate.o_pos)
+        to_blit ;
+      let t =
+        { t with
+          d=
+            { t.d with
+              RFC1951_deflate.o_pos= t.d.RFC1951_deflate.o_pos + to_blit
+            ; RFC1951_deflate.write= t.d.RFC1951_deflate.write + to_blit } }
+      in
+      if off + to_blit = len then k src dst t
+      else
+        Flush
+          { t with
+            z=
+              ctor (fun src dst t ->
+                  (go [@tailcall]) ~ctor (off + to_blit) src dst t ) }
+    in
+    go ~ctor 0 src dst t
+
+  let align ~ctor k src dst t =
+    (* XXX: we ensure than [hold] can not store more than 2 bytes. *)
+    if t.d.RFC1951_deflate.bits > 8 then
+      let k src dst t =
+        k src dst
+          { t with
+            d= {t.d with RFC1951_deflate.hold= 0; RFC1951_deflate.bits= 0} }
+      in
+      put_short_lsb ~ctor t.d.RFC1951_deflate.hold k src dst t
+    else if t.d.RFC1951_deflate.bits > 0 then
+      let k src dst t =
+        k src dst
+          { t with
+            d= {t.d with RFC1951_deflate.hold= 0; RFC1951_deflate.bits= 0} }
+      in
+      put_byte ~ctor t.d.RFC1951_deflate.hold k src dst t
+    else
+      k src dst
+        {t with d= {t.d with RFC1951_deflate.hold= 0; RFC1951_deflate.bits= 0}}
+
+  let put_short_msb ~ctor short k src dst t =
+    let put_byte = put_byte ~ctor in
+    (put_byte ((short lsr 8) land 0xFF) @@ put_byte (short land 0xFF) k)
+      src dst t
+
+  let digest_crc16_byte byte crc16 =
```

Comment:
`digest_crc16_byte` should have this signature: `int -> Crc32.t -> Crc32.t` instead this signature `int -> Crc32.t option -> Crc32.t option`. Then, to use it, it's more idiomatic to use something like `Option.map` and write:

```ocaml
let crc = Option.map digest_byte t.crc in
````